libtenzir/builtins/formats/json.cpp

```diff
@@ -728,40 +888,78 @@ class ndjson_parser final : public parser_base {
       co_return;
     }
     auto& doc = maybe_doc.value_unsafe();
-    auto [action, slices] = this->handle_selector(doc, json_line, state);
-    switch (action) {
-      case parser_action::parse:
-        break;
-      case parser_action::skip:
+    if (precise_) {
+      auto maybe_event = json_to_data(val.value_unsafe(), raw_);
+      if (maybe_event.error()) {
+        // TODO: Extra info?
+        diagnostic::warning("{}", simdjson::error_message(maybe_event.error()))
+          .note("at line {}", lines_processed_)
+          .emit(ctrl_.diagnostics());
         co_return;
-      case parser_action::yield:
-        TENZIR_ASSERT(slices);
+      }
+      auto event = std::move(maybe_event).value_unsafe();
+      if (not caf::holds_alternative<record>(event)) {
+        diagnostic::warning("skipping non-record JSON value: {}", event)
+          .note("at line {}", lines_processed_)
+          .emit(ctrl_.diagnostics());
+        co_return;
+      }
+      signature_.clear();
+      append_signature(event, signature_);
+      auto it = state.precise_map.find(signature_);
+      if (it == state.precise_map.end()) {
+        // TODO: We should eventually garbage collect this.
+        auto index = state.entries.size();
+        state.entries.emplace_back("tenzir.json", std::nullopt);
+        it = state.precise_map.emplace_hint(it, signature_, index);
```

Comment:
I would consider this a follow-up.

Reply:
I think it's clear that we want to have a more generic implementation for this that we can also use for Fluent Bit, CEF, and LEEF. We should do that for that implementaiton.

Reply:
Alright, resolving then. (Also, see tenzir/issues#1809).